plugin/cache/handler.go

```diff
@@ -37,25 +39,22 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 
 		if c.prefetch > 0 {
 			i.Freq.Update(c.duration, now)
-		}
-
-		pct := 100
-		if i.origTTL != 0 { // you'll never know
-			pct = int(float64(ttl) / float64(i.origTTL) * 100)
-		}
 
-		if c.prefetch > 0 && i.Freq.Hits() > c.prefetch && pct < c.percentage {
-			// When prefetching we loose the item i, and with it the frequency
-			// that we've gathered sofar. See we copy the frequencies info back
-			// into the new item that was stored in the cache.
-			prr := &ResponseWriter{ResponseWriter: w, Cache: c, prefetch: true}
-			plugin.NextOrFailure(c.Name(), c.Next, ctx, prr, r)
-
-			if i1, _ := c.get(now, qname, qtype, do); i1 != nil {
-				i1.Freq.Reset(now, i.Freq.Hits())
+			threshold := int(math.Ceil(float64(c.percentage) / 100 * float64(i.origTTL)))
+			if i.Freq.Hits() > c.prefetch && ttl <= threshold {
+				go func() {
```

Comment:
Why does this start another goroutine?

Reply:
See the commit message and pr description.